single-pool/js/packages/client/src/instructions.ts

```diff
@@ -0,0 +1,327 @@
+      ReadonlyAccount<typeof STAKE_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+type WithdrawStakeInstruction = IInstruction<typeof SINGLE_POOL_PROGRAM_ID> &
+  IInstructionWithAccounts<
+    [
+      ReadonlyAccount<PoolAddress>,
+      WritableAccount<PoolStakeAddress>,
+      WritableAccount<PoolMintAddress>,
+      ReadonlyAccount<PoolStakeAuthorityAddress>,
+      ReadonlyAccount<PoolMintAuthorityAddress>,
+      WritableAccount<Base58EncodedAddress>, // user stake
+      WritableAccount<Base58EncodedAddress>, // user token
+      ReadonlyAccount<typeof SYSVAR_CLOCK_ID>,
+      ReadonlyAccount<typeof TOKEN_PROGRAM_ID>,
+      ReadonlyAccount<typeof STAKE_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+type CreateTokenMetadataInstruction = IInstruction<typeof SINGLE_POOL_PROGRAM_ID> &
+  IInstructionWithAccounts<
+    [
+      ReadonlyAccount<PoolAddress>,
+      ReadonlyAccount<PoolMintAddress>,
+      ReadonlyAccount<PoolMintAuthorityAddress>,
+      ReadonlyAccount<PoolMplAuthorityAddress>,
+      WritableSignerAccount<Base58EncodedAddress>, // mpl payer
+      WritableAccount<Base58EncodedAddress>, // mpl account
+      ReadonlyAccount<typeof MPL_METADATA_PROGRAM_ID>,
+      ReadonlyAccount<typeof SYSTEM_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+type UpdateTokenMetadataInstruction = IInstruction<typeof SINGLE_POOL_PROGRAM_ID> &
+  IInstructionWithAccounts<
+    [
+      ReadonlyAccount<VoteAccountAddress>,
+      ReadonlyAccount<PoolAddress>,
+      ReadonlyAccount<PoolMplAuthorityAddress>,
+      ReadonlySignerAccount<Base58EncodedAddress>, // authorized withdrawer
+      WritableAccount<Base58EncodedAddress>, // mpl account
+      ReadonlyAccount<typeof MPL_METADATA_PROGRAM_ID>,
+    ]
+  > &
+  IInstructionWithData<Uint8Array>;
+
+enum SinglePoolInstructionType {
+  InitializePool,
+  DepositStake,
+  WithdrawStake,
+  CreateTokenMetadata,
+  UpdateTokenMetadata,
+}
+
+export class SinglePoolInstruction {
+  static async initializePool(voteAccount: VoteAccountAddress): Promise<InitializePoolInstruction> {
+    const programAddress = SINGLE_POOL_PROGRAM_ID;
+    const pool = await findPoolAddress(programAddress, voteAccount);
+
+    const data = new Uint8Array([SinglePoolInstructionType.InitializePool]);
+
+    return {
+      data,
+      accounts: [
+        { address: voteAccount, role: AccountRole.READONLY },
+        { address: pool, role: AccountRole.WRITABLE },
+        { address: await findPoolStakeAddress(programAddress, pool), role: AccountRole.WRITABLE },
+        { address: await findPoolMintAddress(programAddress, pool), role: AccountRole.WRITABLE },
+        {
+          address: await findPoolStakeAuthorityAddress(programAddress, pool),
+          role: AccountRole.READONLY,
+        },
+        {
+          address: await findPoolMintAuthorityAddress(programAddress, pool),
+          role: AccountRole.READONLY,
+        },
```

Comment:
Serial awaits! The way the interpreter treats these is:

```
const _1 = await findPoolStakeAddress(programAddress, pool), role: AccountRole.WRITABLE };
const _2 = await /* ... */;
const _3 = await /* ... */;
const _4 = await /* ... */;

return {
    data,
    accounts: [
        { address: voteAccount, role: AccountRole.READONLY },
        { address: pool, role: AccountRole.WRITABLE },
        { address: _1, role: AccountRole.WRITABLE },
        { address: _2, role: AccountRole.WRITABLE },
        {address: _3, role: AccountRole.READONLY},
        {address: _4, role: AccountRole.READONLY},
        /* ... */
    ].
}
```

When what you want is to parallelize the work like this:

```suggestion
  static async initializePool(voteAccount: VoteAccountAddress): Promise<InitializePoolInstruction> {
    const programAddress = SINGLE_POOL_PROGRAM_ID;
    const [poolAddress, stakeAddress, mintAddress, stakeAuthority, mintAuthority] = await Promise.all([
      findPoolAddress(programAddress, voteAccount),
      findPoolStakeAddress(programAddress, pool),
      findPoolMintAddress(programAddress, pool),
      findPoolStakeAuthorityAddress(programAddress, pool),
      findPoolMintAuthorityAddress(programAddress, pool),
    ]);

    const data = new Uint8Array([SinglePoolInstructionType.InitializePool]);

    return {
      data,
      accounts: [
        { address: voteAccount, role: AccountRole.READONLY },
        { address: poolAddress, role: AccountRole.WRITABLE },
        { address: stakeAddress, role: AccountRole.WRITABLE },
        { address: mintAddress, role: AccountRole.WRITABLE },
        { address: stakeAuthority, role: AccountRole.READONLY },
        { address: mintAuthority, role: AccountRole.READONLY },
```